File: src/test/java/im/tox/tox4j/DhtNodeSelector.java
package im.tox.tox4j;

import im.tox.tox4j.annotations.NotNull;
import im.tox.tox4j.core.ToxCore;
import im.tox.tox4j.exceptions.ToxException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.InetAddress;
import java.net.Socket;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assume.assumeNotNull;

public final class DhtNodeSelector {

    private static final Logger logger = LoggerFactory.getLogger(DhtNodeSelector.class);
    private static DhtNode selectedNode = ToxCoreTestBase.nodeCandidates[1];

    @NotNull
    public synchronized DhtNode node(ToxFactory factory) {
        if (selectedNode != null) {
            return selectedNode;
        }

        logger.info("Looking for a working bootstrap node");
        for (DhtNode node : ToxCoreTestBase.nodeCandidates) {
            logger.info("Trying to establish a TCP connection to {}", node.ipv4);
            try (Socket socket = new Socket(InetAddress.getByName(node.ipv4), node.port)) {
                assumeNotNull(socket.getInputStream());
            } catch (IOException e) {
                logger.info("TCP connection failed ({})", e.getMessage());
                continue;
            }

            DhtNode found = null;
            boolean success = false;
            for (boolean udpEnabled : new boolean[]{ true, false }) {
                logger.info("Trying to bootstrap with {} using {}", node.ipv4, udpEnabled ? "UDP" : "TCP");
                try (ToxCore tox = factory.newTox(true, udpEnabled)) {
                    // Simple listener for the connected event.
                    ConnectedListener status = new ConnectedListener();
                    tox.callbackConnectionStatus(status);

                    tox.bootstrap(node.ipv4, node.port, node.dhtId);
                    long startTime = System.currentTimeMillis();
                    final long TIMEOUT = 10000;

                    success = false;
                    while (startTime + TIMEOUT > System.currentTimeMillis()) {
                        tox.iteration();
                        Thread.sleep(tox.iterationInterval());

                        if (status.isConnected()) {
                            logger.info("Bootstrapped successfully using {}", udpEnabled ? "UDP" : "TCP");
                            found = node;
                            success = true;
                            break;
                        }
                    }
                    if (!success) {
                        logger.info("Unable to bootstrap with {}", udpEnabled ? "UDP" : "TCP");
                        break;
                    }
                } catch (ToxException | InterruptedException e) {
                    throw new RuntimeException(e);
                }
            }

            if (success) {
                assertNotNull(found);
                return selectedNode = found;
            }
        }

        throw new RuntimeException("No working bootstrap node found");
    }

}

File: src/test/java/im/tox/tox4j/DhtNodeSelector.scala
package im.tox.tox4j

import java.io.IOException
import java.net.{ InetAddress, Socket }

import com.typesafe.scalalogging.Logger
import im.tox.tox4j.core.{ ToxCore, ToxCoreFactory, ToxOptions }
import org.junit.Assume.assumeNotNull
import org.junit.AssumptionViolatedException
import org.slf4j.LoggerFactory

object DhtNodeSelector {

  private val logger = Logger(LoggerFactory.getLogger(this.getClass))
  private var selectedNode: Option[DhtNode] = Some(ToxCoreTestBase.nodeCandidates(1))

  private def tryConnect(node: DhtNode) = {
    var socket: Socket = null
    try {
      socket = new Socket(InetAddress.getByName(node.ipv4), node.udpPort)
      assumeNotNull(socket.getInputStream)
      Some(node)
    } catch {
      case e: IOException =>
        logger.info(s"TCP connection failed (${e.getMessage})")
        None
    } finally {
      if (socket != null) {
        socket.close()
      }
    }
  }

  private def tryBootstrap(factory: (Boolean, Boolean) => ToxCore, node: DhtNode, udpEnabled: Boolean) = {
    val protocol = if (udpEnabled) "UDP" else "TCP"
    val port = if (udpEnabled) node.udpPort else node.tcpPort
    logger.info(s"Trying to bootstrap with ${node.ipv4}:$port using $protocol")

    val tox = factory(true, udpEnabled)

    try {
      val status = new ConnectedListener
      tox.callbackConnectionStatus(status)
      tox.bootstrap(node.ipv4, port, node.dhtId)

      // Try bootstrapping for 10 seconds.
      (0 to 10000 / tox.iterationInterval()) find { _ =>
        tox.iteration()
        Thread.sleep(tox.iterationInterval)
        status.isConnected
      } match {
        case Some(time) =>
          logger.info(s"Bootstrapped successfully after ${time * tox.iterationInterval()}ms using $protocol")
          Some(node)
        case None =>
          logger.info(s"Unable to bootstrap with $protocol")
          None
      }
    } finally {
      tox.close()
    }
  }

  private def findNode(factory: (Boolean, Boolean) => ToxCore): DhtNode = {
    DhtNodeSelector.selectedNode match {
      case Some(node) => node
      case None =>
        logger.info("Looking for a working bootstrap node")

        DhtNodeSelector.selectedNode = ToxCoreTestBase.nodeCandidates find { node =>
          logger.info(s"Trying to establish a TCP connection to ${node.ipv4}")

          (for {
            node <- tryConnect(node)
            node <- tryBootstrap(factory, node, udpEnabled = true)
            node <- tryBootstrap(factory, node, udpEnabled = false)
          } yield node).isDefined
        }

        DhtNodeSelector.selectedNode.getOrElse(
          throw new AssumptionViolatedException("No viable nodes for bootstrap found; cannot test")
        )
    }
  }

  def node: DhtNode = findNode({
    (ipv6Enabled: Boolean, udpEnabled: Boolean) =>
      ToxCoreFactory(new ToxOptions(ipv6Enabled, udpEnabled), null)
  })

}
